### ISSUE TITLE
Implement elementNotToBeCovered for ExpectedConditions library 

### ISSUE THREAD

**Issue by hankduan** [angular/protractor#2313]
See https://github.com/angular/protractor/issues/2139

**Comment by psg99001** [comment]
Really need this. It would solve a big problem

**Comment by evilaliv3** [comment]
This is exaclty our issue in globaleaks dealing with application loader and asyncronous operations.

**Comment by jlin412** [comment]
+1

**Comment by Paulyte** [comment]
Really need this for pop ups and overlays that appear on top of the page, the ExpectedConditions.elementToBeClickable does not help in this case...

**Comment by pavelpp** [comment]
really need this fix as well

**Comment by smearedyam** [comment]
Yes this! I am constantly running into _"Element is not clickable at point (x,y). Other element would receive the click:"_ errors after scrolling and trying to click an element.. but only like ~14% of the time.. frustrating.  Even waiting for elementVisible or elementToBeClickable doesn't always work.

**Comment by wlsf82** [comment]
+1

**Comment by stheikki** [comment]
+1

**Comment by RepoCorp** [comment]
+1

**Comment by beniamin-kis** [comment]
I'd like to see this functionality released ASAP. This issue is so annoying. Is there any workaround for it?

**Comment by bobbyg603** [comment]
+1

**Comment by terodox** [comment]
+1

**Comment by danieljin** [comment]
+1

**Comment by spedy** [comment]
+1

**Comment by jlin412** [comment]
@sjelin you mention that this could be done in another thread, right?
```
We would address this using document.elementFromPoint and element.getBoundingClientRect pretty easily actually. elementFromPoint isn't totally standard at this point, though it appears to be supported by all browsers. The consistency issue is real though. Donno if we should leave this the way it is or "fix" it
```

**Comment by latobibor** [comment]
+1000000000

**Comment by Nicolai82** [comment]
+999

**Comment by ArunBabuVijayanath** [comment]
+2000000

**Comment by ArunBabuVijayanath** [comment]
+2000000

**Comment by anujchandran** [comment]
+999

**Comment by bryceeller2** [comment]
+1

**Comment by magleahy** [comment]
+1

**Comment by Nuri7** [comment]
+1

**Comment by cdodge89** [comment]
+1

**Comment by davidwickman** [comment]
Any update on this?
For my use case, I'm using the following to scroll the element into view, but I know this isn't a solution for all use cases such as modals and such. 

```javascript
var el = $('.myElement');
browser.executeScript('arguments[0].scrollIntoView()', el.getWebElement());
```

**Comment by kseans8** [comment]
We've developed a workaround for now. We identify the element (```id="spinner"```) that is "blocking" the element that we actually want to click and simply just wait for the blocking element to be invisible like so:
```broswer.wait(EC.invisibilityOf($(‘#spinner’)), 5000)```

**Comment by ditkin** [comment]
`EC.invisibilityOf` isn't always enough. Sometimes Protractor reports an element as being covered by another element that shouldn't be invisible.

For instance, in one of my tests, a link is reported as being covered by its parent div, and `invisibilityOf` will wait forever. Seconding the desire for `toNotBeCoveredBy`

**Comment by wlsf82** [comment]
I have solved such issue in one case with the following:
```
"use strict";

function elementWithAttributeHasNotValue(htmlElement, attribute, value) {
    return htmlElement.getAttribute(attribute).then((elementAttribute) => {
        return !elementAttribute.includes(value);
    });
}

class Helper {
    waitForAngularInRoomAnimationToBeDone() {
        const inRoomElement = element(by.className("in-room"));

        browser.wait(elementWithAttributeHasNotValue(inRoomElement, "class", "ng-animate"), browser.params.DEFAULT_TIMEOUT_MS);
    }
}

module.exports = Helper;
```
The name of the method and of the constant are specific from my use case, but I think the rest could be used in a more generic way.

**Comment by MihailSeykov** [comment]
The thing is that in my situation, when I look at the screenshot of the error, which is capturing the state of the browser screen at the moment of the error, everything looks fine and the element I want to click is visible, but Protractor is still reporting that the element is not clickable, because another would receive the click and waiting for element to be visible or enabled doesn't work as the element is already visible on the screen and is enabled. I have this issue in Chrome and it is not appearing every time, only like 20% of the time. I tried the code above, but it does not work for me.

**Comment by conor-mac-aoidh** [comment]
@MihailSeykov I have found that if there is an animation, the screenshot might be taken towards the end of the animation and so it looks like the element is not visible when in fact it is. You can use `EC.and` to wait for `invisibilityOf` multiple elements that might be covering that element

**Comment by wlsf82** [comment]
Sometimes the element will still be visible, but in another place after the animation

**Comment by bradrust** [comment]
I appreciate all the hard work developers give to open source projects (I try and do when I can).   

But,... I really can't see why this isn't the most important default/block/problem to be resolved.  I can't write any tests whatsoever without a bunch of sleeps to wait for items to be hidden or visible in the dom.    I have tried many combinations of wait for invisibility or wait for visibility... but can't write tests that aren't hacky hacky hack hacks.  

I don't understand why everyone doesn't have this visibility problem.

**Comment by Syslog2905** [comment]
This is more than 2 years old and still not fixed??! It is really IMPORTANT for testing w/t hacks, please fix ASAP!

**Comment by klun40** [comment]
Is there any progress on that :)
I though, I finally had a cure for my problem

**Comment by MihailSeykov** [comment]
Can we please have some feedback on this. Is someone looking at it or it will never be fixed? It is causing a lot of problems for a lot of people...

**Comment by wswebcreation** [comment]
As far as I know it's not on the roadmap. But to be able to try to find a solution for this issue it would be nice if we have an example project and example Protractor Code to "proof" where the real issue is. Is that possible?

I'm also reading a lot in the comments about the problem with scrolling to the element and that in some cases it still fails. This could have to do with the fact that scrolling itself also takes time, but having said that, an example project would be very nice!

Who can provide us with that?

**Comment by anneq** [comment]
I'd like to add that for me, this too is the no.1 problem I have with Selenium. Very annoying that you have to build in waits, etc. to account for animations.

**Comment by alllx** [comment]
I have wrote small check utility method, keep in mind it will click on element immediately when it become clickable:
```
import { ElementFinder, promise } from 'protractor';
export let testHelpers = {
  isClickable(el: ElementFinder): promise.Promise<boolean> {
    return new promise.Promise(resolve => {
      let interval = setInterval(() => {
        el.click().then(() => {
          clearInterval(interval);
          setTimeout(() => {
            resolve(true);
          }, 500);
        }, () => { });
      }, 100);
    });
  }
}
```
In your test code:
```
import { testHelpers } from '../src/core/e2e/helpers';
describe('App', () => {
  it('should do something', () {
    let btn = $('.cls');
    browser.wait(testHelpers.isClickable(btn), 3000);
  });
});
```

**Comment by MihailSeykov** [comment]
Can we please get a response from someone involved in the project finally, after more than 2 years? It is costing a lot of us a lot of time trying to find a workaround for this failings and the tests are not stable!!

**Comment by nmfernandes** [comment]
My workaround has been, before clicking on the element, removing from the DOM the element that is covering the element that I want to click.
`await browser.executeScript("arguments[0].remove();", coverElement);`

**Comment by whyboris** [comment]
We're almost at 3 years in ... is this requested feature not solvable from within _Protractor_?
Our tests have `sleep` all over to account for the animations that block element clickability 😞

**Comment by kahan002** [comment]
@nmfernandes  I think it might be a little less dramatic but still work to use arguments[0].style.visibility='hidden';. But in any case, your idea was helpful. Thanks for sharing.

**Comment by Skidush** [comment]
Almost 4 years in, can we please have this? It's very inconvenient to create short sleeps between interactions just to work around this issue. In addition, the `sleeps` accumulate a lot of time which I (and I believe many people also) would like to avoid.

**Comment by PauloGoncalvesBH** [comment]
Does anyone have a code where I can simulate this problem? If possible pass me the github repository to only clone and execute.

**Comment by GitSage** [comment]
I highly doubt that the Protractor team can do anything about this. To me it looks like an issue with either Selenium or Chromedriver.

**Comment by GYuriy** [comment]
+ 1

**Comment by sjelin** [comment]
Unassigning myself because I no longer work for Google

**Comment by smickovskid** [comment]
I created a helper function usinc async / await for myself that takes 1 or 2 arguments as input, `elem` and `retryCount`. What it does it tries to click on an element every .5 sec and if any error is thrown it tries to click again until the retryCount runs out or the button is clicked.

`export const clickOnElementWithRetry = async (elem: ElementFinder, retryCount: number = 5) => {

    while (retryCount) {
        try {
            await elem.click();
            retryCount = 0;
        }
        catch (e) {
            if(retryCount === 0)
                throw "Couldn't click element";
            await browser.sleep(500);
            retryCount--;   
        }
    }

}`

In the spec file or wherever  you want to use it:
`  
it('Clicks on some element', async () => {
    const someButton= element(by.id(`someElementId`));
    clickOnElementWithRetry(someButton)
  })
`

**Comment by Praveen-kumar-test** [comment]
Please provide some function for clicking an element in ExpectedConditions library, and we are suffering by using browser.sleep() in our projects.
